docs/design/proposals/stateless-csi-plugin.md

```diff
@@ -0,0 +1,507 @@
+1. We should be able to generate *ImageName* given a *volume_id*
+  - This is needed in most NS operations, that receive only the *volume_id*
+    and conditionally the *volume_context* and *publish_context*
+  - *volume_context* and *publish_context* are only cached, and not
+    essentially persisted, hence we cannot rely on payload from the same
+    to complete the transformation
+  - This transformation is needed by the CS specifically for DeleteVolume,
+    DeleteSnapshot and ControllerExpandVolume operations
+1. We should generate the same *ImageName*, when a Create request comes in with
+  the same *volume_name*
+  - CreateVolume may retry with the same *volume_name* on timeouts of the
+    request, and hence we need to generate the same *ImageName* for a given
+    *volume_name*
+1. We should be able to distinguish *ImageName*s that come from an instance of
+  the CSI plugin
+  - When dealing with a common Ceph cluster shared across CO systems, we need
+    the *ImageName* to be distinguishable, to avoid collisions and also to
+    be able support/aid operations like ListVolumes/Snapshots
+  - The same also applies when dealing with different CSI plugin instance in
+    the same CO system, using the same Ceph cluster
+
+#### Using Ceph OMaps for idempotent RPCs
+
+To preserve the idempotent CreateVolume RPC requirement, across *ImageName* for
+a given *volume_name*, it is proposed to use a per-CSI instance Ceph omap named
+`csi.volumes.[csi-id]`, to store keys named `csi.volume.[volume_name]` with the
+CO generated *volume_name*.
+
+The value stored in `csi.volume.[volume_name]` is the *ImageName* that is to be
+used for the requested *volume_name*.
+
+A create volume request would hence operate as follows,
+
+- Use CO passed *volume_name* to generate the required key name
+- Check if key exists and read the value of the same, if it does
+- If key exists and value points to an image that satisfies the create request
+  return with success
+- If not, generate an *ImageName* and store the key and value into the omap
+- Proceed with image creation operations as required
+
+Racy key creations/accesses are prevented by locking on the *volume_name* that
+the create request carries, thus ensuring that only one such transaction for
+the key named with the *volume_name* is in flight at any given time.
+
+One other advantage of using this omap is, list operations can be performed
+more easily using the omap, rather than listing all images in the pool or the FS.
+
+NOTE: For RBD plugin the RADOS objects and omaps would be stored in the same
+pool as the image
+
+NOTE: For CephFS plugin the RADOS objects would be stored in the metadata
+pool of the fsname passed to the create request. With further restrictions
+that the objects would be stored in a special `csicephfs` namespace on the
+metadata pool.
+
+#### Naming scheme
+
+ImageName format: `csi-vol-[uuid]`
+
+volume_id format: `[version]-[cluster-id]-[pool-id|fscid]-[uuid]`
```

Comment:
[version], version of what?

Reply:
Version of the volume-id encoding format, helps adapt to future changes to the encoding format (for example adding a namespace ID when supporting using pool namespaces to create RBD images)